pkg/providers/cloudstack/cloudstack.go

```diff
@@ -1057,10 +1083,28 @@ func (p *cloudstackProvider) MachineConfigs(_ *cluster.Spec) []providers.Machine
 	return configs
 }
 
-func (p *cloudstackProvider) UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec) (bool, error) {
+func (p *cloudstackProvider) UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec, cluster *types.Cluster) (bool, error) {
 	newV, oldV := newSpec.VersionsBundle.CloudStack, currentSpec.VersionsBundle.CloudStack
 
-	return newV.ClusterAPIController.ImageDigest != oldV.ClusterAPIController.ImageDigest, nil
+	if newV.ClusterAPIController.ImageDigest != oldV.ClusterAPIController.ImageDigest {
```

Comment:
I'm not sure this is needed
The capi components upgrade us separate from the cluster upgrade
Only components that need a full cluster upgrade should be considered in this method

Reply:
I don't think these are capi components - this is the CAPC controller image. If we're replacing the CAPC controller image that should cause an upgrade right? This is the same behavior we had previously.

Reply:
Ahh good points.

I consider the CAPC controller image part of the capi components (or any other provider controller for that matter).

And we don't need a cluster upgrade just to upgrade the CAPC controller. We manage those component upgrades separately and they can be performed in place without updating the cluster definition.

I see now that code was already there. I don't think it's necessary and in fact it can trigger unnecessary upgrades. I would vote to remove it.

Reply:
Discussed offline - vsphere's components make sense to validate here because they're using a Cluster Resource Set. However, since CloudStack doesn't, this logic is not necessary here